File: src/calculator.cpp
#include "calculator.h"
#include <iostream>
#include <stack>
#include <cmath>
#include <cctype>
#include <stdexcept>
#include <sstream>
#include <map>

using namespace std;

// 检查运算符优先级
int Calculator::getPriority(char op) {
    if (op == '+' || op == '-') return 1;
    if (op == '*' || op == '/') return 2;
    return 0;
}

// 执行基础二元运算
double Calculator::applyOperator(double a, double b, char op) {
    switch (op) {
        case '+': return a + b;
        case '-': return a - b;
        case '*': return a * b;
        case '/': 
            if (b == 0) throw runtime_error("数学错误: 除数不能为零");
            return a / b;
        default: throw runtime_error("不支持的运算符");
    }
}

// 处理单个数字的读取
double Calculator::parseNumber(const string& expr, size_t& index) {
    string numStr;
    while (index < expr.length() && (isdigit(expr[index]) || expr[index] == '.')) {
        numStr += expr[index];
        index++;
    }
    index--; // 回退一位，因为外层循环会再前进一位
    
    try {
        return stod(numStr);
    } catch (...) {
        throw runtime_error("无效数字: " + numStr);
    }
}

// 核心求值函数
double Calculator::evalExpression(const string& expr) {
    stack<double> values;
    stack<char> operators;
    
    for (size_t i = 0; i < expr.length(); i++) {
        if (expr[i] == ' ') continue; // 跳过空格

        if (expr[i] == '(') {
            operators.push('(');
        }
        else if (expr[i] == ')') {
            while (!operators.empty() && operators.top() != '(') {
                double b = values.top(); values.pop();
                double a = values.top(); values.pop();
                char op = operators.top(); operators.pop();
                values.push(applyOperator(a, b, op));
            }
            if (!operators.empty()) operators.pop(); // 弹出 '('
        }
        else if (isdigit(expr[i])) {
            values.push(parseNumber(expr, i));
        }
        else if (expr[i] == '+' || expr[i] == '-' || expr[i] == '*' || expr[i] == '/') {
            // 处理负号（一元运算符）
            if (expr[i] == '-' && (i == 0 || expr[i-1] == '(' || 
                (i>0 && (expr[i-1] == '+' || expr[i-1] == '-' || expr[i-1] == '*' || expr[i-1] == '/')))) {
                values.push(0.0); // 将 -x 转换为 0-x
            }
            // 根据优先级计算
            while (!operators.empty() && getPriority(operators.top()) >= getPriority(expr[i])) {
                double b = values.top(); values.pop();
                double a = values.top(); values.pop();
                char op = operators.top(); operators.pop();
                values.push(applyOperator(a, b, op));
            }
            operators.push(expr[i]);
        }
        else {
            throw runtime_error("错误: 无效字符 '" + string(1, expr[i]) + "' 在表达式中");
        }
    }

    // 处理栈中剩余运算符
    while (!operators.empty()) {
        double b = values.top(); values.pop();
        double a = values.top(); values.pop();
        char op = operators.top(); operators.pop();
        values.push(applyOperator(a, b, op));
    }

    return values.top();
}

// 全局计算函数
string calculate(const string& expr) {
    Calculator calc;
    try {
        double result = calc.evalExpression(expr);
        
        // 处理浮点数精度问题，避免显示 -0.0
        if (abs(result) < 1e-10) result = 0.0;
        
        // 将结果转换为字符串
        stringstream ss;
        ss.precision(10); // 设置精度
        ss << result;
        return ss.str();
    } catch (const exception& e) {
        return "错误: " + string(e.what());
    }
}

File: calculator.js
// calculator.js
// Browser-friendly JS port of calculator.ts

function getPriority(op) {
  if (op === '+' || op === '-') return 1;
  if (op === '*' || op === '/') return 2;
  if (op === '^') return 3;
  return 0;
}

function applyOperator(a, b, op) {
  switch (op) {
    case '+': return a + b;
    case '-': return a - b;
    case '*': return a * b;
    case '/':
      if (Math.abs(b) < Number.EPSILON) throw new Error('数学错误: 除数不能为零');
      return a / b;
    case '^': return Math.pow(a, b);
    default: throw new Error('不支持的运算符');
  }
}

function isDigit(ch) { return ch >= '0' && ch <= '9'; }

function parseNumber(expr, index) {
  let numStr = '';
  const len = expr.length;
  while (index < len && (isDigit(expr[index]) || expr[index] === '.')) {
    numStr += expr[index];
    index++;
  }
  index--;
  if (numStr.length === 0) throw new Error('无效数字: ' + numStr);
  const value = Number(numStr);
  if (isNaN(value)) throw new Error('无效数字: ' + numStr);
  return { value: value, newIndex: index };
}

function popValue(values) {
  if (values.length === 0) throw new Error('数学错误: 操作数不足');
  return values.pop();
}

function evalExpression(expr) {
  const values = [];
  const operators = [];
  const s = expr;
  const len = s.length;

  for (let i = 0; i < len; i++) {
    const ch = s[i];
    if (ch === ' ') continue;
    if (ch === '(') {
      operators.push('(');
    } else if (ch === ')') {
      while (operators.length > 0 && operators[operators.length - 1] !== '(') {
        const b = popValue(values);
        const a = popValue(values);
        const op = operators.pop();
        values.push(applyOperator(a, b, op));
      }
      if (operators.length > 0) operators.pop();
    } else if (isDigit(ch)) {
      const parsed = parseNumber(s, i);
      values.push(parsed.value);
      i = parsed.newIndex;
    } else if (ch === '+' || ch === '-' || ch === '*' || ch === '/' || ch === '^') {
      if (ch === '-' && (i === 0 || s[i-1] === '(' || (i > 0 && (s[i-1] === '+' || s[i-1] === '-' || s[i-1] === '*' || s[i-1] === '/')))) {
        values.push(0.0);
      }
      while (operators.length > 0 && getPriority(operators[operators.length - 1]) >= getPriority(ch)) {
        const b = popValue(values);
        const a = popValue(values);
        const op = operators.pop();
        values.push(applyOperator(a, b, op));
      }
      operators.push(ch);
    } else {
      throw new Error("错误: 无效字符 '" + ch + "' 在表达式中");
    }
  }

  while (operators.length > 0) {
    const b = popValue(values);
    const a = popValue(values);
    const op = operators.pop();
    values.push(applyOperator(a, b, op));
  }

  if (values.length === 0) return 0;
  return values[values.length - 1];
}

function formatNumber(n, significantDigits) {
  if (!isFinite(n)) return String(n);
  let s = n.toPrecision(significantDigits);
  if (s.indexOf('.') >= 0) {
    s = s.replace(/(?:\.0+|0+)$/,'');
    s = s.replace(/\.$/, '');
  }
  return s;
}

export function calculate(expr) {
  try {
    const result = evalExpression(expr);
    const absRes = Math.abs(result);
    const res = (absRes < 1e-10) ? 0.0 : result;
    const out = formatNumber(res, 10);
    return out;
  } catch (e) {
    return '错误: ' + e.message;
  }
}

// expose for convenience
if (typeof window !== 'undefined') window.calculate = calculate;
